configure entity descriptor based on preset value

flyinghermit opened the pull request:
If entity descriptor is not provided when creating a SAML IdP service provider, we used to first try and fetch the entity descriptor from given Entity ID endpoint. If that would fail, we would generate it ourselves. Now, with the `preset` types, we consider preset value to decide whether to go through fetch and gen process or directly generate entity descriptor. As part of this PR, we now directly generate the entity descriptor for the configured `gcp-workforce` preset type. 


This PR also adds the following:
- `Get` func for for the `Preset` field of `SAMLIdPServiceProviderV1` type. 
- Emit error log for the failed entity descriptor generator.

flyinghermit committed:
Get and Set Preset funcs

flyinghermit committed:
configure entity descriptor based on preset value

smallinsky commented on api/types/saml_idp_service_provider.go:
```diff
@@ -129,6 +129,10 @@ type SAMLIdPServiceProvider interface {
 	SetACSURL(string)
+	// GetPreset returns the Preset.
+	GetPreset() string
+	// SetPreset sets the Preset.
+	SetPreset(string)
```
This seems to be not used and only source of preset is yaml/json marshaling.

flyinghermit replied:
Good point, removed in https://github.com/gravitational/teleport/pull/39385/commits/a7788f0f7f6e9890ff705ce6c2f73f11d8a55b2d

smallinsky commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -112,15 +113,9 @@ func (s *SAMLIdPServiceProviderService) GetSAMLIdPServiceProvider(ctx context.Co
-					sp.GetEntityID(), sp.GetACSURL(), err)
-			}
+		if err := s.configureEntityDescriptorPerPreset(sp); err != nil {
+			return trace.Wrap(trace.BadParameter("failed to configure entity descriptor with the given entity_id %q and acs_url %q: %v",
+				sp.GetEntityID(), sp.GetACSURL(), err))
```
I don't thing that you need to use `trace.Wrap` to additionally wrap the `trace.BadParameter`
```suggestion
			return trace.BadParameter("failed to configure entity descriptor with the given entity_id %q and acs_url %q: %v",
				sp.GetEntityID(), sp.GetACSURL(), err)
```

flyinghermit replied:
yaa wrapping it below like you suggested in https://github.com/gravitational/teleport/pull/39385#discussion_r1525921803 seems like more proper place.

smallinsky commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
+func (s *SAMLIdPServiceProviderService) configureEntityDescriptorPerPreset(sp types.SAMLIdPServiceProvider) error {
+	switch sp.GetPreset() {
+	case preset.GCPWorkforce:
+		return s.generateAndSetEntityDescriptor(sp)
```
```suggestion
		return trace.Wrap(s.generateAndSetEntityDescriptor(sp))
```

smallinsky commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
+			s.log.Debugf("Failed to fetch entity descriptor from %q. %v.", sp.GetEntityID(), err)
+			// We aren't interested in checking error type as any occurrence of error
+			// mean entity descriptor was not set.
+			return s.generateAndSetEntityDescriptor(sp)
```
```suggestion
			return trace.Wrap(s.generateAndSetEntityDescriptor(sp))
```

smallinsky commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -243,6 +238,25 @@ func (s *SAMLIdPServiceProviderService) ensureEntityIDIsUnique(ctx context.Conte
+		// fetchAndSetEntityDescriptor is expected to return error if it fails
+		// to fetch a valid entity descriptor.
+		if err := s.fetchAndSetEntityDescriptor(sp); err != nil {
+			s.log.Debugf("Failed to fetch entity descriptor from %q. %v.", sp.GetEntityID(), err)
```
typo `.`  -> `:` ? 
```suggestion
			s.log.Debugf("Failed to fetch entity descriptor from %q: %v.", sp.GetEntityID(), err)
```

flyinghermit committed:
address review comment:
- remove unused SetPreset()
- wrap error while returning configureEntityDescriptorPerPreset
- add error log

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/edgen-per-preset

flyinghermit committed:
fix typo: use semicolon in place of '.'

flyinghermit commented:
Friendly ping @ibeckermayer

flyinghermit committed:
use '%w' instead of '%v' for ErrorF error formatting
